File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = -1001;
    
    public int getMaxPathSum(TreeNode node) {
        //Base Case
        if(node == null) return 0;
        //left and right Recursive calls which will return maxPathSum through respective child nodes
        int l = getMaxPathSum(node.left);
        int r = getMaxPathSum(node.right);
        int maxPathTillMyNode = Math.max(Math.max(l,r) + node.val,node.val);
        int maxPathForMyNode = Math.max(maxPathTillMyNode,l+r+node.val);
        //maximize if this is the global maximum path sum;
        ans = Math.max(ans,maxPathForMyNode);
        return maxPathTillMyNode;
    }
    public int maxPathSum(TreeNode root) {
        getMaxPathSum(root);
        return ans;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans =-1001;
    int getMaxPath(TreeNode* node) {
        //Base Case
        if(!node) return 0;
        //left and right Recursive calls which will return maxPathSum through respective child nodes
        int l = getMaxPath(node->left);
        int r = getMaxPath(node->right);
        //prepare return val (MPS for my parent which passes from any left or right chilf to me)
        int maxPathSumTillMe = max(max(l,r)+node->val,node->val);
        int maxPathValueForMyNode = max(maxPathSumTillMe,l+r+node->val);
        //maximize if this is the global maximum path sum;
        ans = max(ans,maxPathValueForMyNode);
        return maxPathSumTillMe;
    }
        
    int maxPathSum(TreeNode* root) {
        if(root==NULL) return 0;
        cout<<getMaxPath(root);
        return ans;
    }
};